rllib/evaluation/env_runner_v2.py

```diff
@@ -1115,6 +1115,17 @@ def _try_find_policy_again(eval_data: AgentConnectorDataType):
             input_dict = _batch_inference_sample_batches(
                 [d.data.sample_batch for d in eval_data]
             )
+            keys_to_delete = []
+            for key, value in input_dict.items():
+                if not (
+                    np.issubdtype(value.dtype, np.number)
+                    or np.issubdtype(value.dtype, np.bool_)
+                ):
+                    keys_to_delete.append(key)
+            for key in keys_to_delete:
+                del input_dict[key]
+            if "agent_index" in input_dict:
+                del input_dict["agent_index"]
```

Comment:
What if agent index is a legit input to the network?
-> Since it's in our default view requirements, we make it part of the default input dict in dynamic tf policies.
Does this fix still work?

Reply:
custom models may need agent_index. i.e. Policy mapping always maps everything to a single policy. But then the policy figures out how to infuse information from all the agents based on their ids. also we should not delete the keys that don't fit this condition, we should simply not convert them to tensors. How can we achieve that?

Reply:
Also we need VERY VERY VERY clear explanation comments on why we are doing this or any other solution for that matter.

Reply:
ok fair enough on your first point. On your second, the simple list collector's agent_collector was being used for policy inference. That isn't happening here anymore.

Reply:
We have supposedly have replaced that with connector's agent_collector and episode_v2's agent_collector? together they should have played the same role as simple_list collector's agent_collector. Why is that relevant here?

Reply:
https://sourcegraph.com/github.com/ray-project/ray/-/blob/rllib/evaluation/sampler.py?L1206

this line was being invoked before. It is the reason that string agent ids were converted into integers. 
There appears to be no way previously that we would pass a feed dict to build_compute_actions that has non tf tensors e.g. "infos" was never passed to this operation

Reply:
the only solution here is to either delete the keys or to convert them into integers/floats

Reply:
converting them makes sense for the agent_ids but doesn't make sense in the general case.

Reply:
> https://sourcegraph.com/github.com/ray-project/ray/-/blob/rllib/evaluation/sampler.py?L1206
> this line was being invoked before. It is the reason that string agent ids were converted into integers.

This function should be exactly replaced by build_for_inference(). I still don't understand the root cause of the difference in behavior? what is in the previous `get_inference_input_dict` that is different from `build_for_inference` that is treating agent_index differently?

Reply:
Stupid question:
Don't we support nested structures for e.g. the observations here? So these would be filtered out here b/c they are dicts/tuples and not numeric?
Also, what if the model (e.g. PyTorch) expects something in the infos?
How do we know value even has a `dtype` property?

Reply:
What's bad about hashing Agent id into an int again? It's basically a somewhat cleaner solution than what we had before in episode v1 and has little implications.

Reply:
Users never had control over episode's agent IDs and so far I've not heard that it causes trouble.

Reply:
will recircle on this hash's idea later Artur. Good idea.